File: thz-lang-engine/src/tipos.ts
/* ============================================================
 * THZ-LANG — Sistema de Tipos Canônico (análise semântica)
 * ============================================================ */

export enum CategoriaTipo {
  PRIMITIVO = 'PRIMITIVO',
  INTEIRO = 'INTEIRO',
  DECIMAL = 'DECIMAL',
  MONETARIO = 'MONETARIO',
  FATIA = 'FATIA',
  REGISTRO = 'REGISTRO',
  ENUMERACAO = 'ENUMERACAO',
  RESULTADO = 'RESULTADO'
}

export interface TipoThz {
  /** Nome verbatim como escrito no fonte (ex.: 'DECIMAL(12,4)'). */
  nome: string;
  categoria: CategoriaTipo;
  /** Casas decimais para DECIMAL/MONETARIO. */
  escala?: number;
  /** Código ISO 4217 para MONETARIO. */
  moeda?: string;
  /** Tipo interno para FATIA[T] e canal de sucesso de RESULTADO[T,E]. */
  interno?: string;
  /** Canal de erro de RESULTADO[T,E]. */
  internoErro?: string;
}

/** Marcador especial para literais inteiros: cabem em inteiro ou decimal. */
export const TIPO_LITERAL_INTEIRO: TipoThz = { nome: '<literal-inteiro>', categoria: CategoriaTipo.PRIMITIVO };

const PRIMITIVOS: Readonly<Record<string, TipoThz>> = Object.freeze({
  TEXTO: { nome: 'TEXTO', categoria: CategoriaTipo.PRIMITIVO },
  LOGICO: { nome: 'LOGICO', categoria: CategoriaTipo.PRIMITIVO },
  UUID: { nome: 'UUID', categoria: CategoriaTipo.PRIMITIVO },
  DATA: { nome: 'DATA', categoria: CategoriaTipo.PRIMITIVO },
  DATA_HORA: { nome: 'DATA_HORA', categoria: CategoriaTipo.PRIMITIVO }
});

const INTEIROS: Readonly<Record<string, TipoThz>> = Object.freeze({
  NATURAL8: { nome: 'NATURAL8', categoria: CategoriaTipo.INTEIRO },
  NATURAL16: { nome: 'NATURAL16', categoria: CategoriaTipo.INTEIRO },
  NATURAL32: { nome: 'NATURAL32', categoria: CategoriaTipo.INTEIRO },
  NATURAL64: { nome: 'NATURAL64', categoria: CategoriaTipo.INTEIRO },
  INTEIRO8: { nome: 'INTEIRO8', categoria: CategoriaTipo.INTEIRO },
  INTEIRO16: { nome: 'INTEIRO16', categoria: CategoriaTipo.INTEIRO },
  INTEIRO32: { nome: 'INTEIRO32', categoria: CategoriaTipo.INTEIRO },
  INTEIRO64: { nome: 'INTEIRO64', categoria: CategoriaTipo.INTEIRO }
});

/** Registro completo de nomes primitivos reconhecidos pela análise. */
export const TIPOS_PRIMITIVOS: Readonly<Record<string, TipoThz>> = Object.freeze({
  ...PRIMITIVOS,
  ...INTEIROS
});

/**
 * Interpreta um nome de tipo verbatim do fonte.
 * Retorna undefined quando o nome não corresponde a nenhum tipo conhecido.
 */
export function analisarNomeTipo(nomeVerbatim: string): TipoThz | undefined {
  const primitivo = TIPOS_PRIMITIVOS[nomeVerbatim];
  if (primitivo) return primitivo;

  let casamento = /^DECIMAL\s*\(\s*\d+\s*,\s*(\d+)\s*\)$/.exec(nomeVerbatim);
  if (casamento) {
    return { nome: nomeVerbatim.replace(/\s+/g, ''), categoria: CategoriaTipo.DECIMAL, escala: Number.parseInt(casamento[1], 10) };
  }

  casamento = /^MONETARIO\s*\(\s*"?([A-Z]{3})"?\s*\)$/.exec(nomeVerbatim);
  if (casamento) {
    return { nome: nomeVerbatim.replace(/\s+/g, ''), categoria: CategoriaTipo.MONETARIO, moeda: casamento[1] };
  }

  casamento = /^RESULTADO\s*\[\s*([^,\]]+?)\s*,\s*([^,\]]+?)\s*\]$/.exec(nomeVerbatim);
  if (casamento) {
    return {
      nome: nomeVerbatim.replace(/\s+/g, ''),
      categoria: CategoriaTipo.RESULTADO,
      interno: casamento[1].trim(),
      internoErro: casamento[2].trim()
    };
  }

  casamento = /^FATIA\s*\[\s*(\w+)\s*\]$/.exec(nomeVerbatim);
  if (casamento) {
    return { nome: nomeVerbatim.replace(/\s+/g, ''), categoria: CategoriaTipo.FATIA, interno: casamento[1] };
  }

  // Estruturas declaradas no programa são resolvidas pelo analisador.
  return undefined;
}

export function ehInteiro(t?: TipoThz): boolean {
  return t?.categoria === CategoriaTipo.INTEIRO || t === TIPO_LITERAL_INTEIRO;
}

export function ehNumerico(t?: TipoThz): boolean {
  return ehInteiro(t) || t?.categoria === CategoriaTipo.DECIMAL;
}

/**
 * Compatibilidade de atribuição/comparação.
 * Regras v2.2:
 *  - nomes idênticos são sempre compatíveis;
 *  - inteiros (e literais inteiros) convertem implicitamente para DECIMAL;
 *  - MONETARIO nunca mistura com numéricos nem com outras moedas;
 *  - FATIA/REGISTRO exigem identidade nominal.
 * Tipos desconhecidos (undefined) retornam true para não cascatear erros já reportados.
 */
export function saoCompativeis(origem?: TipoThz, destino?: TipoThz): boolean {
  if (!origem || !destino) return true;
  if (origem.nome === destino.nome) return true;
  if (ehInteiro(origem) && ehInteiro(destino)) return true;
  if (ehNumerico(origem) && destino.categoria === CategoriaTipo.DECIMAL) return true;

  return false;
}

/** Descrição legível para mensagens de erro. */
export function descrever(t?: TipoThz): string {
  return t ? "'" + t.nome + "'" : '<desconhecido>';
}

File: JVM/thz-core-jvm/src/main/java/thz/lang/semantico/Tipos.java
package thz.lang.semantico;

import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class Tipos {
    public static final TipoThz TIPO_LITERAL_INTEIRO = new TipoThz("<literal-inteiro>", CategoriaTipo.PRIMITIVO);

    private static final Map<String, TipoThz> PRIMITIVOS = Map.of(
        "TEXTO", new TipoThz("TEXTO", CategoriaTipo.PRIMITIVO),
        "LOGICO", new TipoThz("LOGICO", CategoriaTipo.PRIMITIVO),
        "UUID", new TipoThz("UUID", CategoriaTipo.PRIMITIVO),
        "DATA", new TipoThz("DATA", CategoriaTipo.PRIMITIVO),
        "DATA_HORA", new TipoThz("DATA_HORA", CategoriaTipo.PRIMITIVO)
    );
    private static final Map<String, TipoThz> INTEIROS = Map.of(
        "NATURAL8", new TipoThz("NATURAL8", CategoriaTipo.INTEIRO),
        "NATURAL16", new TipoThz("NATURAL16", CategoriaTipo.INTEIRO),
        "NATURAL32", new TipoThz("NATURAL32", CategoriaTipo.INTEIRO),
        "NATURAL64", new TipoThz("NATURAL64", CategoriaTipo.INTEIRO),
        "INTEIRO8", new TipoThz("INTEIRO8", CategoriaTipo.INTEIRO),
        "INTEIRO16", new TipoThz("INTEIRO16", CategoriaTipo.INTEIRO),
        "INTEIRO32", new TipoThz("INTEIRO32", CategoriaTipo.INTEIRO),
        "INTEIRO64", new TipoThz("INTEIRO64", CategoriaTipo.INTEIRO)
    );

    public static final Map<String, TipoThz> TIPOS_PRIMITIVOS;
    static {
        var m = new java.util.HashMap<String, TipoThz>();
        m.putAll(PRIMITIVOS);
        m.putAll(INTEIROS);
        TIPOS_PRIMITIVOS = Map.copyOf(m);
    }

    private Tipos() {}

    public static TipoThz analisarNomeTipo(String verbatim) {
        TipoThz prim = TIPOS_PRIMITIVOS.get(verbatim);
        if (prim != null) return prim;
        if ("DECIMAL".equals(verbatim)) return new TipoThz("DECIMAL", CategoriaTipo.DECIMAL, 4, null, null, null);
        if ("INTEIRO".equals(verbatim)) return new TipoThz("INTEIRO64", CategoriaTipo.INTEIRO);
        if ("NATURAL".equals(verbatim)) return new TipoThz("NATURAL64", CategoriaTipo.INTEIRO);
        if ("MONETARIO".equals(verbatim) || "DINHEIRO".equals(verbatim)) return new TipoThz("MONETARIO", CategoriaTipo.MONETARIO, null, "BRL", null, null);
        if ("REGISTRO".equals(verbatim) || "RECORD".equals(verbatim)) return new TipoThz("REGISTRO", CategoriaTipo.REGISTRO, null, null, "REGISTRO", null);
        if ("MAPA".equals(verbatim) || "MAP".equals(verbatim)) return new TipoThz("MAPA", CategoriaTipo.MAPA, null, null, "MAPA", null);
        Matcher m;
        m = Pattern.compile("^DECIMAL\\s*\\(\\s*\\d+\\s*,\\s*(\\d+)\\s*\\)$").matcher(verbatim);
        if (m.matches()) return new TipoThz(verbatim.replaceAll("\\s+", ""), CategoriaTipo.DECIMAL, Integer.parseInt(m.group(1)), null, null, null);
        m = Pattern.compile("^MONETARIO\\s*\\(\\s*\"?([A-Z]{3})\"?\\s*\\)$").matcher(verbatim);
        if (m.matches()) return new TipoThz(verbatim.replaceAll("\\s+", ""), CategoriaTipo.MONETARIO, null, m.group(1), null, null);
        m = Pattern.compile("^RESULTADO\\s*\\[\\s*([^,\\]]+?)\\s*,\\s*([^,\\]]+?)\\s*\\]$").matcher(verbatim);
        if (m.matches()) return new TipoThz(verbatim.replaceAll("\\s+", ""), CategoriaTipo.RESULTADO, null, null, m.group(1).trim(), m.group(2).trim());
        m = Pattern.compile("^FATIA\\s*\\[\\s*(\\w+)\\s*\\]$").matcher(verbatim);
        if (m.matches()) return new TipoThz(verbatim.replaceAll("\\s+", ""), CategoriaTipo.FATIA, null, null, m.group(1), null);
        return null;
    }

    public static boolean ehInteiro(TipoThz t) {
        return t != null && (t.categoria() == CategoriaTipo.INTEIRO || t == TIPO_LITERAL_INTEIRO);
    }
    public static boolean ehNumerico(TipoThz t) {
        return ehInteiro(t) || (t != null && t.categoria() == CategoriaTipo.DECIMAL);
    }
    public static boolean saoCompativeis(TipoThz origem, TipoThz destino) {
        if (origem == null || destino == null) return true;
        if (origem.nome().equals(destino.nome())) return true;
        if (ehInteiro(origem) && ehInteiro(destino)) return true;
        if (ehNumerico(origem) && destino.categoria() == CategoriaTipo.DECIMAL) return true;
        if (origem.categoria() == CategoriaTipo.REGISTRO && destino.categoria() == CategoriaTipo.REGISTRO) return true;
        if (origem.categoria() == CategoriaTipo.MAPA && destino.categoria() == CategoriaTipo.MAPA) return true;
        return false;
    }
    public static String descrever(TipoThz t) {
        return t != null ? "'" + t.nome() + "'" : "<desconhecido>";
    }
}
